src/data/quaternion.lean

```diff
@@ -0,0 +1,318 @@
+  right_inv := conj_conj,
+  add := conj_add,
+  smul := conj_smul }
+
+variable {R}
+
+@[simp] lemma conj_lin_equiv_symm : (conj_lin_equiv R).symm = conj_lin_equiv R := rfl
+
+@[simp] lemma coe_conj_lin_equiv : ⇑(conj_lin_equiv R) = conj := rfl
+
+@[simp] lemma conj_zero : conj (0:ℍ[R]) = 0 := (conj_lin_equiv R).map_zero
+
+@[simp] lemma conj_neg : (-a).conj = -a.conj := (conj_lin_equiv R).map_neg a
+
+@[simp] lemma conj_sub : (a - b).conj = a.conj - b.conj := (conj_lin_equiv R).map_sub a b
+
+open opposite
+
+variable (R)
+
+/-- Quaternion conjugate as a `ring_equiv` to the opposite ring. -/
+def conj_ring_equiv : ℍ[R] ≃+* (ℍ[R]ᵒᵖ) :=
+{ to_fun := op ∘ conj,
+  inv_fun := conj ∘ unop,
+  map_mul' := λ x y, by simp,
+ .. (conj_lin_equiv R).to_add_equiv.trans (add_equiv.op_equiv ℍ[R]) }
+
+variable {R}
+
+@[simp] lemma coe_conj_ring_equiv : ⇑(conj_ring_equiv R) = op ∘ conj := rfl
+
+variable (R)
+
+/-- Square of the norm. -/
+def norm_sq : ℍ[R] →* R :=
+{ to_fun := λ a, (a * a.conj).re,
+  map_one' := by rw [conj_one, one_mul, has_one_re],
+  map_mul' := λ x y, coe_injective $ by conv_lhs { rw [← mul_conj_eq_coe, conj_mul, mul_assoc,
+    ← mul_assoc y, y.mul_conj_eq_coe, coe_commutes, ← mul_assoc, x.mul_conj_eq_coe, ← coe_mul] } }
+
+variable {R}
+
+lemma norm_sq_def : norm_sq R a = (a * a.conj).re := rfl
+
+lemma norm_sq_def' : norm_sq R a = a.1^2 + a.2^2 + a.3^2 + a.4^2 :=
+by simp only [norm_sq_def, pow_two, ← neg_mul_eq_mul_neg, sub_neg_eq_add] with quaternion_projs
+
+lemma norm_sq_coe : norm_sq R x = x^2 := by rw [norm_sq_def, conj_coe, ← coe_mul, coe_re, pow_two]
+
+lemma norm_sq_zero : norm_sq R 0 = 0 :=
+by simp only [norm_sq_def', pow_two, mul_zero, add_zero] with quaternion_projs
+
+@[simp] lemma norm_sq_neg : norm_sq R (-a) = norm_sq R a :=
+by simp only [norm_sq_def, conj_neg, neg_mul_neg]
+
+lemma self_mul_conj : a * a.conj = norm_sq R a := by rw [mul_conj_eq_coe, norm_sq_def]
+
+lemma conj_mul_self : a.conj * a = norm_sq R a := by rw [← a.commute_self_conj.eq, self_mul_conj]
+
+end quaternion
+
+namespace quaternion
+
+variables {R : Type*}
+
+section linear_ordered_comm_ring
+
+variables [linear_ordered_comm_ring R] {a : ℍ[R]}
+
+@[simp] lemma norm_sq_eq_zero : norm_sq R a = 0 ↔ a = 0 :=
+begin
+  refine ⟨λ h, _, λ h, h.symm ▸ norm_sq_zero⟩,
+  rw [norm_sq_def', add_eq_zero_iff_eq_zero_of_nonneg, add_eq_zero_iff_eq_zero_of_nonneg,
+    add_eq_zero_iff_eq_zero_of_nonneg] at h,
+  exact ext a 0 (pow_eq_zero h.1.1.1) (pow_eq_zero h.1.1.2) (pow_eq_zero h.1.2) (pow_eq_zero h.2),
+  all_goals { apply_rules [pow_two_nonneg, add_nonneg] }
+end
+
+lemma norm_sq_ne_zero : norm_sq R a ≠ 0 ↔ a ≠ 0 := not_congr norm_sq_eq_zero
+
+@[simp] lemma norm_sq_nonneg : 0 ≤ norm_sq R a :=
+by { rw norm_sq_def', apply_rules [pow_two_nonneg, add_nonneg] }
+
+@[simp] lemma norm_sq_le_zero : norm_sq R a ≤ 0 ↔ a = 0 :=
+by simpa only [le_antisymm_iff, norm_sq_nonneg, and_true] using @norm_sq_eq_zero _ _ a
+
+instance : domain ℍ[R] :=
```

Comment:
This doesn't depend on the order structure, right? I might want to apply this to a quaternion algebra over a number field (or its ring of integers)...

Reply:
Can you prove `norm_sq R a = 0 ↔ a = 0` without the order structure?

Reply:
Yes, of course you are right. Sorry. For number fields you need to have a real embedding. So they don't have a canonical order structure, and we'll have to do more manual work anyways... Let's leave it like it is.